### PR TITLE
Update CPUManager stored state semantics

### PR DESCRIPTION
**Pull request by klueska**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test

/kind feature

> /kind flake

**What this PR does / why we need it**:
This PR updates the information that is stored as part of the CPUManager state. Instead of storing the CPUManager state keyed off of the `ContainerID`, we now store it keyed off of the combination of `PodUID`and `ContainerName`.  This change is necessary moving forward as we move to a model where we assign CPUs to containers at pod admit time, rather than container creation time.

**Does this PR introduce a user-facing change?**:

```release-note
The underlying format of the `CPUManager` state file has changed. Upgrades should be seamless, but any third-party tools that rely on reading the previous format need to be updated.
```

### CODE REVIEW

**Review by mattjmcnaughton — COMMENTED** [review]
I've been a little detached from the reviews on the CPUManager internals, so I'll leave technical specifics around this change to someone else :)

But, I do have one question. From reading the PR description, it looked like we plan to merge this diff to master before the subsequent necessary diff you mentioned is approved and ready to merge? If yes, that makes me a little nervous. If for some reason, the subsequent diff takes longer, it'd be unfortunate (I think? perhaps you can share more about the ramifications) to have the first diff in 1.17 but not the second.

**Comment by klueska** [comment]
@mattjmcnaughton This patch should only be merged together with the patch to support upgrades. I'm working on this right now and will add it as a follow-on commit in this same PR once it's done, and update the description accordingly.

**Comment by mattjmcnaughton** [comment]
> @mattjmcnaughton This patch should only be merged together with the patch to support upgrades. I'm working on this right now and will add it as a follow-on commit in this same PR once it's done, and update the description accordingly.

Gotcha, thanks @klueska ! Mind adding a `[WIP]` prefix to your PR title, so its clear to reviewers and doesn't get accidentally merged? Thanks!

**Comment by derekwaynecarr** [comment]
@klueska this appears to take commits from the #84196 that is also in progress.  updating the title to WIP pending the core PRs merging first.

**Comment by klueska** [comment]
@ConnorDoyle @derekwaynecarr 

This PR has now been rebased on top of master (well technically on top of #84525 which I'm hoping will merge soon).

However, we can't call this PR complete until we have a migration path of the old state format (keyed off of containerID) to the new state format (keyed off of podUID, containerName).

I'm curious what your thoughts are for how to ensure state stored by 1.16 kubelets will be readable by 1.17 kubelets, transformed to the new semantics, and then written back in the new format.

I imagine there must be some a sort of precedent on best practices for this, but my (admittedly limited) search for it hasn't turned up much.

**Comment by cbf123** [comment]
In OpenStack nova they had a mechanism that was basically opportunistic upgrades...that is, the code understood how to read both the old format and the new format, and when it read in the data in the old format it would convert the data to the new format and write it back.

I'm not sure it's needed here, but there was a separate "online upgrade" stage to ensure that all of the data had been updated to the new format.  This stage needed to be manually run before upgrading to a version that no longer supported the old data format.

Do we need to support fallback in case of a failed upgrade?  If so, it might make sense to use a different filename for the new saved state file, or maybe to write the new data to the state file in a way that the old version could still read.  In a future release you could delete the backwards-compatibility state information.

**Comment by klueska** [comment]
The bigger issue is that we can’t immediately convert from one format to another as a „pre-step“ because we don’t have the pod uid or container name readily available as we are doing the recovery. We can do it lazily (which is probably what I’ll end up doing), but then it gets more complicated with keeping things in sync (not too bad though).

**Comment by klueska** [comment]
Having an upgrade stage supported during kubelet startup would help a lot in this regard. Basically a state restore phase, followed by and upgrade phase before any components begin their normal startup.

As far as I know, no coordinated effort exists at present. All components just do their own checkpoint management as they come online, assuming no dependencies on any other components. Normally this is fine (you should checkpoint everything you need to recover in a component by component basis), but it’s problematic when upgrading to a format, where going forward you need additional info not in the old state file (as is the case here).

**Comment by derekwaynecarr** [comment]
/assign

i will review this today.

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 1]
```diff
@@ -96,6 +97,10 @@ type manager struct {
 	// and the containerID of their containers
 	podStatusProvider status.PodStatusProvider
 
+	// containerMap provides a mapping from (pod, container) -> containerID
```
Comment (derekwaynecarr) [inline pkg/kubelet/cm/cpumanager/cpu_manager.go]: nit: its containerID -> (pod, container) per earlier commits

File: pkg/kubelet/cm/cpumanager/state/checkpoint.go [hunk 2]
```diff
@@ -23,38 +23,87 @@ import (
...
 	DefaultCPUSet string            `json:"defaultCpuSet"`
 	Entries       map[string]string `json:"entries,omitempty"`
 	Checksum      checksum.Checksum `json:"checksum"`
 }
 
+// CPUManagerCheckpointV2 struct is used to store cpu/pod assignments in a checkpoint in v2 format
+type CPUManagerCheckpointV2 struct {
+	PolicyName    string                       `json:"policyName"`
+	DefaultCPUSet string                       `json:"defaultCpuSet"`
+	Entries       map[string]map[string]string `json:"entries,omitempty"`
```
Comment (derekwaynecarr) [inline pkg/kubelet/cm/cpumanager/state/checkpoint.go]: nit: a map to a map can be cumbersome, we can come back to this later with an actual type.

**Comment by derekwaynecarr** [comment]
the breakdown per commit was really helpful, and the mechanical changes all made sense.

/approve
/lgtm

**Comment by neolit123** [comment]
@klueska i have a suspicion that this PR is making the 1.18 kubelet unstable, resulting in panics.

see this issue for more details:
https://github.com/kubernetes/kubernetes/issues/86676

**Pull request merged**